mac80211: do not enforce start_disabled=1 when staidx>0

msvamp opened the pull request:
I do not really understand the purpose of not letting the AP interface start automatically when there are wireless client (station mode) interfaces present on the same wireless radio.
Maybe there are issues in some drivers that do not let this work properly? Or is it because client connection can disrupt the access point interface connectivity? I don't know.

With this patch, I provide an option for the user to override this default behaviour by explicitly setting `start_disabled` to `0`. The existence of station interfaces will only control the default value of `start_disabled` but the user's preference will always be used if it is explicitly defined.

ptpt52 commented:
mark.

neheb commented on package/kernel/mac80211/files/lib/netifd/wireless/mac80211.sh:
```diff
@@ -511,13 +511,13 @@ mac80211_hostapd_setup_bss() {
 	}
-	[ "$staidx" -gt 0 -o "$start_disabled" -eq 1 ] && append hostapd_cfg "start_disabled=1" "$N"
+
+	[ "$staidx" -gt 0 ] && set_default start_disabled 1 || set_default start_disabled 0
```
wonder if shellcheck would warn that A && B || C is not if else . Probably better to use an if statement just in case.

msvamp replied:
Interesting, what's shellcheck?

I know that it's not if-else but set_default would never return false, I tried it.

neheb replied:
https://www.shellcheck.net/wiki/SC2015

shellcheck is a static analyzer for shell files. These files are not really checked and full of shellcheck warnings.

msvamp replied:
Hmm, is there any work being done about rewriting this in ucode? 😅

neheb replied:
doubt it. Would be an interesting PR.

Ansuel replied:
> Hmm, is there any work being done about rewriting this in ucode? 😅

that would be really funny!

msvamp committed:
mac80211: do not enforce start_disabled=1 when staidx>0

Allow to override "start_disabled" to 0 even if there are station
interfaces present on the radio. If "start_disabled" is not explicitly
defined, the original behaviour of setting it to 1 is retained.
In other words, the existence of station interfaces now only controls
the default value of "start_disabled", but does not enforce it to 1.

Signed-off-by: Manas Sambhus <manas.sambhus+github@gmail.com>

msvamp commented:
@ptpt52 @neheb @Ansuel Can you please review it again and suggest if it looks good now?

msvamp commented:
Ok, I was still on 22.03 and didn't notice that we have actually shifted to ucode-based configuration on master now. Please ignore my proposed changes above, I will go through the new implementation and redo this PR.

Also, I noticed that simply adding `start_disabled=0` is not sufficient anymore to allow AP interface to be running after STA interface disconnects. Something is shutting down the AP interface when it disconnects.
